### PR TITLE
Backend logging framework w/ Azure SQL DB integration POC

### PR DESCRIPTION
**Pull request by swoocn**
Logging record result in Azure SQL DB ->

image

**[see commits]**

**_TODOs before PR merge_**
+ Hide SQL DB credentials; stash into secrets/ environment variables.
+ Log level and logging instances - are we logging what is only necessary and to keep from filling up the tables with just "stuff"?
+ Build VERCEL_ENV variable for Prod in the backend server (API hosting); has this been done @soleil00?
+ Currently, the primary key in the table is the timestamp value; this will eventually be changed to sessionId+timestamp.
+ Will need to share AppDev access to Azure SQL DB client.

### CODE REVIEW

File: backend/src/controllers/productControllers.js [hunk 1]
```diff
@@ -1,6 +1,12 @@
 const Product = require("../models/productModel");
 const Shop = require("../models/shopModel");
-const logger = require("../../logger");
+const initializeLogger = require("../../logger");
+
+let logger;
+
+initializeLogger().then((initializedLogger) => {
+  logger = initializedLogger;
+});
```
Comment (swoocn) [inline backend/src/controllers/productControllers.js:9]: Not really satisfied with this implementation for initializing the logger every time it is needed, but couldn't think of any good workaround as of yet.  Feel free to suggest ways to refactor this.

File: backend/src/controllers/productControllers.js [hunk 2]
```diff
@@ -18,7 +24,7 @@ const addProductToShop = async (req, res) => {
         shop.products.push(newProduct);
         await shop.save();
 
-        logger.info("Product added successfully");
+        logger.info("Product added successfully", newProduct);
```
Comment (swoocn) [inline backend/src/controllers/productControllers.js:27]: Would appreciate thoughts on what we should and should not log in the DB table. 🙃

**Comment by zolcsi** [comment]
@swoocn, me looks good! I have just high level questions. The logs are stored in an azure cloud mssql, what is the reason behind, could we put it into mongodb as well? But if it is a classic sql db though, we could make the primary key the timestamp as it is now, and secondary key the sessionId. So that we can query based efficiently with sessionId. If document based DB, the primary key could be a composite key (sessionid + timestamp) to be able to query sessionID w/wo time ranges

**Review by zolcsi — APPROVED** [review]

**Comment by swoocn** [comment]
> @swoocn, me looks good! I have just high level questions. The logs are stored in an azure cloud mssql, what is the reason behind, could we put it into mongodb as well? But if it is a classic sql db though, we could make the primary key the timestamp as it is now, and secondary key the sessionId. So that we can query based efficiently with sessionId. If document based DB, the primary key could be a composite key (sessionid + timestamp) to be able to query sessionID w/wo time ranges

Great question @zolcsi!
If you recall our last conversation, we agreed with storing logs in a relational DB for stronger data consistency and querying capabilities.  I chatted with Ant who generously provided database resources from his free Azure subscription, so we went with its SQL services.  However, after working with it a bit, I now think a document-based DB i.e., MongoDB may be a better choice as it will help us better manage and scale high volume of logging data.  I can certainly switch the implementation over to MongoDB since I also experimented with that during the process.  Rather than supplementing the SQL DB transport, let me go ahead and  replace with a MongoDB transport.

**Comment by swoocn** [comment]
Changed to Mongo DB logging transport  ->

image

**Comment by zolcsi** [comment]
HI @swoocn, all clear, thank you for the explanation. It's your call, either way it's good, but I also think it's less overhead if we keep stuff at the same place. Regardless, many thanks to @Ant as well for the resources

**Pull request merged**